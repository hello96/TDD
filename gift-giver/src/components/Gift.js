import React, { Component } from 'react';
import {
  Form,
  FormGroup,
  FormControl,
  ControlLabel,
  Button
} from 'react-bootstrap';

export default class Gift extends Component {
  state = {
    person: '',
    present: ''
  };

  setPerson = ({ target: { value: person } }) => {
    this.setState({
      person
    });
  };

  setPresent = ({ target: { value: present } }) => {
    this.setState({
      present
    });
  };

  render() {
    const {
      gift: { id: giftId },
      removeGift
    } = this.props;

    return (
      <>
        <Form>
          <FormGroup>
            <ControlLabel>Person</ControlLabel>
            <FormControl className="input-person" onChange={this.setPerson} />
          </FormGroup>
          <FormGroup>
            <ControlLabel>Present</ControlLabel>
            <FormControl className="input-present" onChange={this.setPresent} />
          </FormGroup>
          <Button
            className="btn-remove btn-danger"
            onClick={() => removeGift(giftId)}
          >
            Remove a gift
          </Button>
        </Form>
      </>
    );
  }
}
