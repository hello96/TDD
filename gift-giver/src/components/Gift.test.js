import React from 'react';
import { shallow } from 'enzyme';
import Gift from './Gift';

describe('Gift', () => {
  const removeGiftMock = jest.fn();
  const gift = shallow(<Gift gift={{ id: 0 }} removeGift={removeGiftMock} />);

  it('should render', () => {
    expect(gift).toMatchSnapshot();
  });

  it('initializes a person and present in `state`', () => {
    expect(gift.state()).toEqual({
      person: '',
      present: ''
    });
  });

  describe('when typing into the preson input', () => {
    const input = gift.find('.input-person');

    beforeEach(() => {
      input.simulate('change', { target: { value: 'Uncle Mle' } });
    });

    afterEach(() => {
      gift.setState({
        person: '',
        present: ''
      });
    });

    it('should update the state', () => {
      expect(gift.state().person).toEqual('Uncle Mle');
    });
  });

  describe('when typing into the present input', () => {
    const input = gift.find('.input-present');

    beforeEach(() => {
      input.simulate('change', { target: { value: 'Croccantino' } });
    });

    afterEach(() => {
      gift.setState({
        person: '',
        present: ''
      });
    });

    it('should update the state', () => {
      expect(gift.state().present).toEqual('Croccantino');
    });
  });

  describe('when clicking the `Remove Gift` button', () => {
    beforeEach(() => {
      gift.find('.btn-remove').simulate('click');
    });

    it('calls the removeGift callback', () => {
      expect(removeGiftMock).toHaveBeenCalledTimes(1);
      expect(removeGiftMock).toHaveBeenCalledWith(0);
    });
  });
});
