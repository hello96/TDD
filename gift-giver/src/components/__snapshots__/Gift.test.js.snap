// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gift should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Gift
    gift={
      Object {
        "id": 0,
      }
    }
    removeGift={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Form
        bsClass="form"
        componentClass="form"
        horizontal={false}
        inline={false}
      >
        <FormGroup
          bsClass="form-group"
        >
          <ControlLabel
            bsClass="control-label"
            srOnly={false}
          >
            Person
          </ControlLabel>
          <FormControl
            bsClass="form-control"
            className="input-person"
            componentClass="input"
            onChange={[Function]}
          />
        </FormGroup>
        <FormGroup
          bsClass="form-group"
        >
          <ControlLabel
            bsClass="control-label"
            srOnly={false}
          >
            Present
          </ControlLabel>
          <FormControl
            bsClass="form-control"
            className="input-present"
            componentClass="input"
            onChange={[Function]}
          />
        </FormGroup>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="default"
          className="btn-remove btn-danger"
          disabled={false}
          onClick={[Function]}
        >
          Remove a gift
        </Button>
      </Form>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bsClass": "form",
        "children": Array [
          <FormGroup
            bsClass="form-group"
          >
            <ControlLabel
              bsClass="control-label"
              srOnly={false}
            >
              Person
            </ControlLabel>
            <FormControl
              bsClass="form-control"
              className="input-person"
              componentClass="input"
              onChange={[Function]}
            />
          </FormGroup>,
          <FormGroup
            bsClass="form-group"
          >
            <ControlLabel
              bsClass="control-label"
              srOnly={false}
            >
              Present
            </ControlLabel>
            <FormControl
              bsClass="form-control"
              className="input-present"
              componentClass="input"
              onChange={[Function]}
            />
          </FormGroup>,
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="btn-remove btn-danger"
            disabled={false}
            onClick={[Function]}
          >
            Remove a gift
          </Button>,
        ],
        "componentClass": "form",
        "horizontal": false,
        "inline": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": Array [
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
              >
                Person
              </ControlLabel>,
              <FormControl
                bsClass="form-control"
                className="input-person"
                componentClass="input"
                onChange={[Function]}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "control-label",
                "children": "Person",
                "srOnly": false,
              },
              "ref": null,
              "rendered": "Person",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "form-control",
                "className": "input-person",
                "componentClass": "input",
                "onChange": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": Array [
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
              >
                Present
              </ControlLabel>,
              <FormControl
                bsClass="form-control"
                className="input-present"
                componentClass="input"
                onChange={[Function]}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "control-label",
                "children": "Present",
                "srOnly": false,
              },
              "ref": null,
              "rendered": "Present",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "form-control",
                "className": "input-present",
                "componentClass": "input",
                "onChange": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "active": false,
            "block": false,
            "bsClass": "btn",
            "bsStyle": "default",
            "children": "Remove a gift",
            "className": "btn-remove btn-danger",
            "disabled": false,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Remove a gift",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Form
          bsClass="form"
          componentClass="form"
          horizontal={false}
          inline={false}
        >
          <FormGroup
            bsClass="form-group"
          >
            <ControlLabel
              bsClass="control-label"
              srOnly={false}
            >
              Person
            </ControlLabel>
            <FormControl
              bsClass="form-control"
              className="input-person"
              componentClass="input"
              onChange={[Function]}
            />
          </FormGroup>
          <FormGroup
            bsClass="form-group"
          >
            <ControlLabel
              bsClass="control-label"
              srOnly={false}
            >
              Present
            </ControlLabel>
            <FormControl
              bsClass="form-control"
              className="input-present"
              componentClass="input"
              onChange={[Function]}
            />
          </FormGroup>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="btn-remove btn-danger"
            disabled={false}
            onClick={[Function]}
          >
            Remove a gift
          </Button>
        </Form>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bsClass": "form",
          "children": Array [
            <FormGroup
              bsClass="form-group"
            >
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
              >
                Person
              </ControlLabel>
              <FormControl
                bsClass="form-control"
                className="input-person"
                componentClass="input"
                onChange={[Function]}
              />
            </FormGroup>,
            <FormGroup
              bsClass="form-group"
            >
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
              >
                Present
              </ControlLabel>
              <FormControl
                bsClass="form-control"
                className="input-present"
                componentClass="input"
                onChange={[Function]}
              />
            </FormGroup>,
            <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="default"
              className="btn-remove btn-danger"
              disabled={false}
              onClick={[Function]}
            >
              Remove a gift
            </Button>,
          ],
          "componentClass": "form",
          "horizontal": false,
          "inline": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": Array [
                <ControlLabel
                  bsClass="control-label"
                  srOnly={false}
                >
                  Person
                </ControlLabel>,
                <FormControl
                  bsClass="form-control"
                  className="input-person"
                  componentClass="input"
                  onChange={[Function]}
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "control-label",
                  "children": "Person",
                  "srOnly": false,
                },
                "ref": null,
                "rendered": "Person",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "form-control",
                  "className": "input-person",
                  "componentClass": "input",
                  "onChange": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": Array [
                <ControlLabel
                  bsClass="control-label"
                  srOnly={false}
                >
                  Present
                </ControlLabel>,
                <FormControl
                  bsClass="form-control"
                  className="input-present"
                  componentClass="input"
                  onChange={[Function]}
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "control-label",
                  "children": "Present",
                  "srOnly": false,
                },
                "ref": null,
                "rendered": "Present",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "form-control",
                  "className": "input-present",
                  "componentClass": "input",
                  "onChange": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "active": false,
              "block": false,
              "bsClass": "btn",
              "bsStyle": "default",
              "children": "Remove a gift",
              "className": "btn-remove btn-danger",
              "disabled": false,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Remove a gift",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
